jina/peapods/pods/compoundpod.py

```diff
@@ -0,0 +1,271 @@
+
+        .. note::
+            If one of the :class:`Pod` fails to start, make sure that all of them
+            are properly closed.
+        """
+        if getattr(self.args, 'noblock_on_start', False):
+            for _args in self.all_args['peas']:
+                _args.noblock_on_start = True
+                self._enter_pea(Pea(_args))
+            for _args in self.all_args['replicas']:
+                _args.noblock_on_start = True
+                _args.polling = PollingType.ALL
+                self._enter_replica(Pod(_args))
+
+            # now rely on higher level to call `wait_start_success`
+            return self
+        else:
+            try:
+                for _args in self.all_args['peas']:
+                    self._enter_pea(Pea(_args))
+                for _args in self.all_args['replicas']:
+                    self._enter_replica(Pod(_args))
+            except:
+                self.close()
+                raise
+            return self
+
+    def wait_start_success(self) -> None:
+        """
+        Block until all pods and peas start successfully.
+        If not successful, it will raise an error hoping the outer function to catch it
+        """
+
+        if not self.args.noblock_on_start:
+            raise ValueError(
+                f'{self.wait_start_success!r} should only be called when `noblock_on_start` is set to True'
+            )
+
+        try:
+            for p in self.peas:
+                p.wait_start_success()
+            for p in self.replica_list:
+                p.wait_start_success()
+        except:
+            self.close()
+            raise
+
+    def _enter_replica(self, replica: 'Pod') -> None:
+        self.replica_list.append(replica)
+        self.enter_context(replica)
+
+    def join(self):
+        """Wait until all pods and peas exit."""
+        try:
+            for p in self.peas:
+                p.join()
+            for p in self.replica_list:
+                p.join()
+        except KeyboardInterrupt:
+            pass
+        finally:
+            self.peas.clear()
+            self.replica_list.clear()
+
+    @property
+    def is_ready(self) -> bool:
+        """
+        Checks if Pod is read.
+        :return: true if the peas and pods are ready to serve requests
+
+        .. note::
+            A Pod is ready when all the Peas it contains are ready
+        """
+        return all(
+            [p.is_ready.is_set() for p in self.peas]
+            + [p.is_ready.is_set() for p in self.replica_list]
+        )
+
+    def _set_after_to_pass(self, args):
+        if PollingType.ANY.is_push:
+            # ONLY reset when it is push
+            args.uses_after = '_pass'
+
+    def rolling_update(self):
+        """
+        Update all pods of this compound pod.
+        """
+        for i in range(len(self.replica_list)):
+            replica = self.replica_list[i]
+            replica.close()
+            del replica
```

Comment:
remove this delete, since it does nothing

Reply:
thanks, I thought I removed it already